File: leetcode/leetcode_py/number-of-recent-calls.py
'''
https://leetcode.com/problems/number-of-recent-calls

1. Queue and pop out of range
time: o(n)
space: o(n)
'''
class RecentCounter:

    def __init__(self):
        self.q = []

    def ping(self, t: int) -> int:
        self.q.append(t)
        while self.q[0] < t - 3000:
            self.q.pop(0)
        return len(self.q)

# Your RecentCounter object will be instantiated and called as such:
# obj = RecentCounter()
# param_1 = obj.ping(t) 

File: leetcode/leetcode_cpp/number-of-recent-calls.cpp
/*
https://leetcode.com/problems/number-of-recent-calls

1. Queue and pop out of range
time: o(n)
space: o(n)
*/
class RecentCounter {
public:
    RecentCounter() {
        
    }
    
    int ping(int t) {
        q.push(t); // logn
        while (q.front() < t - 3000) q.pop(); // n
        return q.size();
    }

    queue<int> q;
};

/**
 * Your RecentCounter object will be instantiated and called as such:
 * RecentCounter* obj = new RecentCounter();
 * int param_1 = obj->ping(t);
 */